probe-rs/src/probe/wlink/mod.rs

```diff
@@ -0,0 +1,606 @@
+use core::fmt;
+use std::{thread::sleep, time::Duration};
+
+use rusb::{Device, UsbContext};
+
+use crate::{
+    architecture::riscv::communication_interface::{RiscvCommunicationInterface, RiscvError},
+    DebugProbe, DebugProbeError, DebugProbeInfo, DebugProbeSelector, DebugProbeType,
+    ProbeCreationError, WireProtocol,
+};
+
+use self::usb_interface::WchLinkUsbDevice;
+use super::JTAGAccess;
+
+mod usb_interface;
+
+const VENDER_ID: u16 = 0x1a86;
+const PRODUCT_ID: u16 = 0x8010;
+
+const DMI_OP_NOP: u8 = 0;
+const DMI_OP_READ: u8 = 1;
+const DMI_OP_WRITE: u8 = 2;
+
+#[derive(Debug)]
+#[repr(u8)]
+pub enum WchLinkVariant {
+    /// WCH-Link-CH549, does not support CH32V00X
+    Ch549 = 1,
+    /// WCH-LinkE-CH32V307,
+    ECh32v307 = 2, // or V305?
+    /// WCH-LinkS-CH32V203 ,
+    SCh32v203 = 3,
+    /// WCH-LinkB,
+    B = 4,
+}
+
+impl fmt::Display for WchLinkVariant {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            WchLinkVariant::Ch549 => write!(f, "WCH-Link-CH549"),
+            WchLinkVariant::ECh32v307 => write!(f, "WCH-LinkE-CH32V307"),
+            WchLinkVariant::SCh32v203 => write!(f, "WCH-LinkS-CH32V203"),
+            WchLinkVariant::B => write!(f, "WCH-LinkB"),
+        }
+    }
+}
+
+#[derive(Debug)]
+#[repr(u8)]
+pub enum RiscvChip {
+    CH32V103 = 0x01,
+    CH573 = 0x02,
+    CH569 = 0x03,
+    CH32V20x = 0x05,
+    CH32V30x = 0x06,
+    CH32V003 = 0x09,
+}
+
+impl RiscvChip {
+    fn from_u8(value: u8) -> Option<Self> {
+        match value {
+            0x01 => Some(RiscvChip::CH32V103),
+            0x02 => Some(RiscvChip::CH573),
+            0x03 => Some(RiscvChip::CH569),
+            0x05 => Some(RiscvChip::CH32V20x),
+            0x06 => Some(RiscvChip::CH32V30x),
+            0x09 => Some(RiscvChip::CH32V003),
+            _ => None,
+        }
+    }
+}
+
+/// WCH-Link device (mod:RV)
+#[derive(Debug)]
+pub(crate) struct WchLink {
+    device: WchLinkUsbDevice,
+    name: String,
+    variant: WchLinkVariant,
+    v_major: u8,
+    v_minor: u8,
+    riscvchip: u8,
+    chip_type: u32,
+    // Hack to support NOP after READ
+    last_dmi_read: Option<(u8, u32, u8)>,
+}
+
+impl WchLink {
+    fn get_version(&mut self) -> Result<(), DebugProbeError> {
+        tracing::debug!("Getting version of WCH-Link...");
+
+        let rxbuf = self
+            .device
+            .write_command(&[0x81, 0x0d, 0x01, 0x01], Duration::from_millis(300))?;
```

Comment:
Please document the format of the commands somewhere. At least with a comment, but having a enum or something would be best.

Reply:
Will do. I need to do some reverse engineering.
These fields are all undocumented. Only a few can be guessed easily.

Reply:
I'm also happy with a comment that it's reverse engineered, then it's clear. No need to invest too much effort.

Reply:
I will add a document link here.